### ISSUE TITLE
Keep upgrading subresource requests to preloaded HSTS hosts

### ISSUE THREAD

**Issue by michaelkleber** [mikewest/strict-navigation-security#1]
One strong vote for keeping subresource upgrades when the host is on the HSTS preload list (as the [doc mentions](https://github.com/mikewest/strict-navigation-security#couldnt-we-safely-apply-hsts-upgrades-to-subresource-requests-to-preloaded-hsts-hosts) alread), where there is no worry about browser-by-browser variation.

My use case: 3rd-party services that Google has been running for many years that are implemented by sites cutting-and-pasting a snippet of Google-provided HTML into their web pages.  Examples include Google Analytics and AdSense.

The HTML blob that these services hand out _now_ loads everything over https, but that wasn't the case ten years ago.  Many sites copy-and-pasted such a blob once and never update it.  HSTS is a way that these embedded snippets can become more secure without a million different web sites to update themselves individually, and it would be a pity to lose that capability.

**Comment by mikewest** [comment]
That is a very reasonable argument. I have aesthetic/philosophical objections to distinguishing between the statically preloaded list and dynamic pins, but I think you're right that the practical implications are pretty compelling.

**Comment by estark37** [comment]
I also feel sad about adding a new dependency on the preload list, because it could make it harder to get rid of one day.

Would it work if we disallowed subresources on http:// pages from writing HSTS, and from reading HSTS state that was set by subresources on https:// pages? That is,
- http:// pages wouldn't be able to read HSTS cookies set on https:// pages (and https:// pages can't read them because of mixed content autoupgrading)
- http:// pages can't write their own HSTS cookies
- Somewhat similar security properties to applying only the preload list, but a little better (dynamic-HSTS subresources on http:// pages would still get upgraded if the user had previously visited them as 1p) and doesn't special-case the preload list.

cc @davidben

**Comment by davidben** [comment]
@michaelkleber I assume this use case is just http sites? https sites already get the mixed content rules independent of HSTS. The question then is how important it is to upgrade those scripts to https, once the cookie mess is sorted. I'm not sure where I land on that one. The containing page is already fully network-controlled, so it's meaningless under the usual https threat model. But some http URLs may be more easily attacked than others, especially if they're a choke point or not, as the FAQ [notes](https://github.com/mikewest/strict-navigation-security#does-this-enable-interesting-network-based-attacks).

@estark37 Is the idea we store separate HSTS entries for "set by top-level navigation" and "set by subresource"? And then http:// pages would only pay attention to the former? And then either way we'd still apply the preload list. So somewhere in between fully partitioning by first-party and the current situation. You can still read cookies set on navigations, though that is more trouble to do. Suppose a site spent some round trips on top-level redirects (maybe over several navigations to hide the latency) to add HSTS pins, then the http pages can read that cookie back.

We could also swing further double-keying dynamic HSTS but still allow the preload list as in your idea. Double-keying makes reasoning about privacy easier, but it makes dynamic pins less effective, so it's more weight on the preload list. (Still less than exclusively using the preload list.) We also have the fuzzy importance of these upgrades to begin with, which helps decrease the impact of all this.

**Comment by estark37** [comment]
> Is the idea we store separate HSTS entries for "set by top-level navigation" and "set by subresource"? And then http:// pages would only pay attention to the former? And then either way we'd still apply the preload list. So somewhere in between fully partitioning by first-party and the current situation.

Yeah, something like that. Another way to put it is that we'd consider the preload list part of the "set by top-level navigation" partition.

> You can still read cookies set on navigations, though that is more trouble to do. Suppose a site spent some round trips on top-level redirects (maybe over several navigations to hide the latency) to add HSTS pins, then the http pages can read that cookie back.

Yeah, I'm not sure how to reason about that. I don't know if I buy that trackers would be willing to do top-level redirects to write a cookie, but not to read one... but I also don't know much about this ecosystem. ¯\_(ツ)_/¯

> We could also swing further double-keying dynamic HSTS but still allow the preload list as in your idea.

I was kind of thinking that we should try to protect against same-site tracking as well as cross-site tracking, that is http://example.com shouldn't be able to set an HSTS cookie even for only itself to read. That is, I'm thinking of it as not just a cross-site tracking mechanism but also a supercookie (defining supercookie as "something that behaves like a cookie but doesn't disappear when you clear cookies/site data"), and we should try to kill off the supercookie as well as the cross-site tracking.

**Comment by davidben** [comment]
> Yeah, I'm not sure how to reason about that. I don't know if I buy that trackers would be willing to do top-level redirects to write a cookie, but not to read one... but I also don't know much about this ecosystem. ¯_(ツ)_/¯

Hrm, good point. Honestly, I'm still very much in the figuring stuff out and making things up stages myself. :-) Top-level navigations and top-level redirects are kind of a headache in general since we end up switching first-party contexts in them. E.g. you can also try to pass identifiers in URLs. It's not clear to me what we should cut where to get a coherent story for how to handle navigation tricks in general.

**Comment by davidben** [comment]
Here's a variation without the navigation tricks: pick a pile of moderately popular, HSTS-asserting, unpreloaded hostnames. If I could learn the subset of those sites that any individual user visits, that would give me a reasonable amount of unpartitioned entropy. So I embed images from each one and try to determine if it was upgraded, either via timing (if the site isn't cooperating) or some other signal (if the site is cooperating, it could make the responses all visible via CORS and just tell me).
